source/h2_decoder.c

```diff
@@ -15,48 +15,642 @@
+    } else {
+        s_decoder_reset_state(decoder);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+static int s_state_fn_frame_headers(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+    (void)decoder;
+    (void)input;
+
+    /* Read padding and/or finish frame */
+    if (decoder->frame_in_progress.payload_len == 0) {
+
+        if (decoder->frame_in_progress.flags & AWS_H2_FRAME_F_END_HEADERS) {
+            DECODER_CALL_VTABLE_STREAM(decoder, on_end_headers);
+        }
+
+        if (decoder->frame_in_progress.flags & AWS_H2_FRAME_F_PADDED) {
+            s_decoder_set_state(decoder, &s_state_padding);
+        } else {
+            s_decoder_reset_state(decoder);
+        }
+    }
+
+    return AWS_OP_ERR;
+}
+static int s_state_fn_frame_priority(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+    (void)input;
+
+    /* No data to process here, we're done! */
+    s_decoder_reset_state(decoder);
+
+    return AWS_OP_SUCCESS;
+}
+static int s_state_fn_frame_rst_stream(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+    (void)decoder;
+    (void)input;
+
+    AWS_FATAL_ASSERT(input->len >= 4);
+
+    uint32_t error_code = 0;
+    bool succ = aws_byte_cursor_read_be32(input, &error_code);
+    AWS_ASSERT(succ);
+    (void)succ;
+
+    DECODER_CALL_VTABLE_STREAM_ARGS(decoder, on_rst_stream, error_code);
+
+    return AWS_OP_SUCCESS;
+}
+static int s_state_fn_frame_settings(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+    (void)decoder;
+    (void)input;
+
+    /* If ack is set, report and abort */
+    if (decoder->frame_in_progress.flags & AWS_H2_FRAME_F_ACK) {
+        DECODER_CALL_VTABLE(decoder, on_settings_ack);
+        s_decoder_reset_state(decoder);
+        return AWS_OP_SUCCESS;
+    }
+
+    /* Otherwise, we need to process blocks of 6 octets at a time */
+    decoder->state.bytes_required = s_setting_block_size;
+
+    /* Truck through the list until we run out of space */
+    while (input->len >= s_setting_block_size) {
+        uint16_t id = 0;
+        uint32_t value = 0;
+
+        bool succ = aws_byte_cursor_read_be16(input, &id);
+        AWS_ASSERT(succ);
+        (void)succ;
+
+        succ = aws_byte_cursor_read_be32(input, &value);
+        AWS_ASSERT(succ);
+        (void)succ;
+
+        DECODER_CALL_VTABLE_ARGS(decoder, on_setting, id, value);
+
+        /* Update payload len */
+        decoder->frame_in_progress.payload_len -= s_setting_block_size;
+    }
+
+    if (decoder->frame_in_progress.payload_len == 0) {
+        /* Huzzah, done with the frame */
+        s_decoder_reset_state(decoder);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+static int s_state_fn_frame_push_promise(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
+    (void)decoder;
+    (void)input;
+    return AWS_OP_ERR;
+}
+static int s_state_fn_frame_ping(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
```

Comment:
> PING frames are not associated with any individual stream. If a PING frame is received with a stream identifier field value other than 0x0, the recipient MUST respond with a connection error (Section 5.4.1) of type PROTOCOL_ERROR.

Do any other states need that check as well?
Or should we check in some common place that non-stream frames don't have stream-id set?

Reply:
I can definitely do a validation pass, do you mind if I do that as a separate PR?

Reply:
Put a TODO somewhere in the code about it